### ISSUE TITLE
[iOS][Database] Returned snapshot does not match web API

### ISSUE THREAD

**Issue by HofmannZ** [invertase/react-native-firebase#100]
Yesterday I stumbled upon an issue with react-native-firebase, where the behavior is different than the Firebase web API.

## Actual behavior
I have a ref with the keys being numbers, like this:
```js
{
  1: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  2: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  3: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
};
```

*Note that it starts with 1 and not with 0.*

When I request the ref with react-native-firebase I get the following snapshot.

```js
[
  null,
    1: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  2: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  3: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
];
```

## Expected behavior
When stored and retrieved with the Firebase web API, I just get the first example back.
I expected that react-native-firebase would return the same object, like so:

```js
{
  1: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  2: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
  3: {
    someKey: 'someValue',
    someOtherKey: 'someOtherValue',
  },
};
```

## Possible issue and notes
I'm not good with C++, nor the native iOS Firebase API, but I think it has to do with some conversion to an array?
Also, it worked fine on Android.

## Versions
- react: 16.0.0-alpha.6
- react-native: 0.44.0
- react-native-firebase: 1.0.2

**Comment by Salakar** [comment]
![image](https://cloud.githubusercontent.com/assets/5347038/25857227/53d297f4-34d0-11e7-934c-9b9c83aefc82.png)

Yep looks like you're right, have done a test in the tests app, can see it erroring. Looking into it.

**Comment by Salakar** [comment]
Hmmm wait a minute.

This actually looks like its as intended by firebase, here's the web sdk's response:

![image](https://cloud.githubusercontent.com/assets/5347038/25857929/5f815f48-34d2-11e7-907e-0800d74dd360.png)

Notice 3 visible items, but the array length is 4 - meaning it also has a null item in there

**Comment by HofmannZ** [comment]
Yeah, I thought It might be something like that, this actually returns an array. In react-native I see it as an object with one item being null.
If it would just return an array it would be fine, but an object with null is weird. 
I would personally like to see an object returned without the null, because it matches the web API. But I can image you might just want to return an array instead of an object.

**Comment by HofmannZ** [comment]
The web API just returns an object and storing arrays in firebase is depreciated.

And this is really weird:
`typeof` Returns 'object':
<img width="397" alt="screen shot 2017-05-09 at 17 50 23" src="https://cloud.githubusercontent.com/assets/17142193/25859900/3160715e-34e0-11e7-98fc-4fa6e22552e6.png">

But actualy is an array:
<img width="295" alt="screen shot 2017-05-09 at 17 50 46" src="https://cloud.githubusercontent.com/assets/17142193/25859906/350e6b58-34e0-11e7-8d37-3ab8b086ff91.png">

And on Firebase, it's an object:
<img width="264" alt="screen shot 2017-05-09 at 17 54 51" src="https://cloud.githubusercontent.com/assets/17142193/25860079/b31b4b10-34e0-11e7-873b-d11870ee951a.png">

Really needs some consitancey...

**Comment by HofmannZ** [comment]
Still a bug, the android API does not match the web and iOS API.

**Comment by HofmannZ** [comment]
Still a bug, the android API does not match the web and iOS API.

**Comment by Ehesp** [comment]
Arrays are objects in JS, which is expected. Will flag it as an Android bug, needs some work doing on the Java->JS transformer.

**Comment by chrisbianca** [comment]
Just pushed a fix for this

**Comment by jsm** [comment]
I'm seeing a difference between iOS and Android still.

iOS returns an object, while android seems to return an Array.

**Comment by Ehesp** [comment]
@jsm what version are you on?

**Comment by jsm** [comment]
I'm on `1.1.0` but it looks like this was patched in `1.0.3` has anything else changed since `1.1.0`?

**Comment by chrisbianca** [comment]
There was a further fix in the 2.0.X and 2.1.X streams which resolves this issue, along with lots of other fixes and features.  You should look to upgrade.

**Comment by jsm** [comment]
Sweet, thanks. What exactly was the fix? Have a link to the commits?
